File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/shell/AppPreferences.kt
package lk.mageride.passenger.shell

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.content.SharedPreferences
import android.content.res.Configuration
import androidx.core.content.edit
import lk.mageride.shared.data.models.Language
import java.util.Locale

/**
 * The handful of answers the app needs **before** it has a session.
 *
 * SCR-PA-002 asks for a language on a handset that has never signed in, and SCR-PA-005 asks for
 * location permission. Neither can be written to `iam.users` at the moment it is given, so they
 * are held here and pushed to the server on the first authenticated call — which is C077's job,
 * not the shell's.
 *
 * **The shell owns this store because the shell reads it from `attachBaseContext`**, before a
 * composition exists and while Koin's graph is reachable only through the global context. It is an
 * interface because a view-model test has no `SharedPreferences` — a local unit test's stub
 * answers a default for every member, which would make a language choice look like it saved and
 * silently do nothing.
 */
internal interface AppPreferences {

    /** The chosen UI language (D-26, AL-26). `null` until SCR-PA-002 has been answered. */
    var language: Language?

    /** Whether the local choice still has to be pushed to `iam.users` after sign-in. */
    var languagePendingSync: Boolean

    /** Whether SCR-PA-005 has been shown and dismissed. Grants themselves are asked of the OS. */
    var locationRationaleAcknowledged: Boolean

    /**
     * SCR-PA-015a's *"remembers last choice"* — the wire value of the last [CallType] used, or
     * `null` before the first call.
     *
     * Held here rather than in the ride's state because it outlives every ride: a passenger who
     * always calls normally should not be asked again on their next trip. Δ C079/C080.
     */
    var lastCallType: String?

    /**
     * Whether the *"your number is visible to the other party"* tooltip has been shown (US-26.5).
     *
     * Once, on the **first** call. AL-48 withdrew number masking, so this is the PDPA transparency
     * that replaced it — and a disclosure shown on every call is a disclosure nobody reads.
     */
    var callNumberNoticeShown: Boolean

    /**
     * SCR-PA-027's *"Default payment"* — the wire value of the chosen `PaymentMethod`, or `null`
     * before the passenger has chosen one (US-22.4, AL-14).
     *
     * Here rather than only on `iam.users` because the column cannot hold every rail this app
     * offers: `DefaultPaymentMethod` is still `[cash, lankaqr, onepay]` and AL-57 replaced `onepay`
     * with the **wallet**, which has no value in that enum. See `PaymentPreference`. Δ C083.
     */
    var defaultPaymentMethod: String?

    /** Whether SCR-PA-002 has been answered — "first-launch only" (D2' §A). */
    val firstRunComplete: Boolean get() = language != null
}

/** [AppPreferences] over a private `SharedPreferences` file. */
internal class AndroidAppPreferences(context: Context) : AppPreferences {

    private val store: SharedPreferences =
        context.applicationContext.getSharedPreferences(FILE, Context.MODE_PRIVATE)

    override var language: Language?
        get() = store.getString(KEY_LANGUAGE, null)?.let(Language::fromWire)
        set(value) = store.edit { putString(KEY_LANGUAGE, value?.wire) }

    override var languagePendingSync: Boolean
        get() = store.getBoolean(KEY_PENDING_SYNC, false)
        set(value) = store.edit { putBoolean(KEY_PENDING_SYNC, value) }

    override var locationRationaleAcknowledged: Boolean
        get() = store.getBoolean(KEY_LOCATION_RATIONALE, false)
        set(value) = store.edit { putBoolean(KEY_LOCATION_RATIONALE, value) }

    override var lastCallType: String?
        get() = store.getString(KEY_LAST_CALL_TYPE, null)
        set(value) = store.edit { putString(KEY_LAST_CALL_TYPE, value) }

    override var callNumberNoticeShown: Boolean
        get() = store.getBoolean(KEY_CALL_NUMBER_NOTICE, false)
        set(value) = store.edit { putBoolean(KEY_CALL_NUMBER_NOTICE, value) }

    override var defaultPaymentMethod: String?
        get() = store.getString(KEY_DEFAULT_PAYMENT, null)
        set(value) = store.edit { putString(KEY_DEFAULT_PAYMENT, value) }

    private companion object {
        const val KEY_LAST_CALL_TYPE = "last_call_type"
        const val KEY_CALL_NUMBER_NOTICE = "call_number_notice_shown"
        const val KEY_DEFAULT_PAYMENT = "default_payment_method"

        // Not the C018 database: `mobile_db_schema.md` §0.4 keeps that file encrypted and opening
        // it is `suspend`, and `attachBaseContext` cannot wait for a Keystore round trip to know
        // which locale to inflate resources in. Nothing here is a secret.
        const val FILE = "passenger_prefs"
        const val KEY_LANGUAGE = "language"
        const val KEY_PENDING_SYNC = "language_pending_sync"
        const val KEY_LOCATION_RATIONALE = "location_rationale_acknowledged"
    }
}

/**
 * Applies the language SCR-PA-002 chose to the resources the app inflates.
 *
 * D-26 makes the language a *user* preference, not the handset's: a passenger on a phone set to
 * English who picks සිංහල must get a Sinhala app, and AL-26 makes Sinhala the **default** rather
 * than a translation of English. Android's per-app locale API (`LocaleManager`) is API 33+, and
 * the URD NFR-22 floor is API 26 — so the base context is wrapped instead, which works on every
 * level this app supports and needs no `appcompat`.
 *
 * The wrap is applied in `MainActivity.attachBaseContext`, which is the only place early enough:
 * by `onCreate` the `Resources` a composable will read have already been resolved.
 */
internal object PassengerLocale {

    /**
     * [base] with the stored language applied, or [base] itself when SCR-PA-002 has not been
     * answered (the handset's own locale is the right answer until the passenger says otherwise).
     */
    fun wrap(base: Context): Context {
        val language = AndroidAppPreferences(base).language ?: return base
        val locale = Locale.forLanguageTag(language.wire)

        // `Locale.setDefault` covers what does not go through `Resources` — number and date
        // formatting in particular, which a fare and a scheduled pickup both depend on.
        Locale.setDefault(locale)

        val configuration = Configuration(base.resources.configuration).apply {
            setLocale(locale)
        }
        return ContextWrapper(base.createConfigurationContext(configuration))
    }
}

/**
 * The `Activity` hosting this context, or `null`.
 *
 * Compose hands a screen a `Context`, and a language change has to reach the one thing that can
 * re-inflate resources — `Activity.recreate()`. Walking the `ContextWrapper` chain is the only way
 * to it from inside a composable.
 */
internal tailrec fun Context.findActivity(): Activity? = when (this) {
    is Activity -> this
    is ContextWrapper -> baseContext.findActivity()
    else -> null
}

File: apps/passenger-ios/PassengerApp/Shell/AppPreferences.swift
import Foundation
import MageRideShared

/// The device-local answers this app keeps outside its database.
///
/// **Everything here is a *device* setting rather than an account one**, and each is here because
/// there is no route that stores it: AL-26 makes the interface language a device-first choice the
/// server write is allowed to lag, SCR-PI-005's rationale is a first-run fact, and C101's default
/// payment rail has no value in `iam.yaml`'s `DefaultPaymentMethod` enum for `wallet` at all (the
/// contract gap C076 recorded, unchanged).
///
/// A protocol with a `UserDefaults` implementation for the reason the Android twin is an interface:
/// a model test needs to set a value rather than reach for a suite, and the seam is what makes the
/// language rules assertable without a running app.
///
/// The keys are `apps/passenger-android/.../shell/AppPreferences.kt`'s, so a passenger who moves
/// between platforms is not the only thing the two apps disagree about. ``lastCallType`` and
/// ``callNumberNoticeShown`` arrived with C098, which is the component that first asks them —
/// SCR-PI-015a is the screen with the memory, exactly as the Info.plist's camera key arrived with the
/// commit that opens a camera. A stored answer nothing asks for is state nobody maintains.
protocol AppPreferences: AnyObject {

    /// SCR-PI-002's answer. `nil` on a first run, which is what makes AL-26's Sinhala default a
    /// *pre-selection* rather than a stored value nobody chose.
    var language: Language? { get set }

    /// Whether the chosen language still has to be written to `iam.users` (AL-26).
    ///
    /// Set when the device is written and the server is not — SCR-PI-002 runs before there is a
    /// session at all, so the first write is always local-only. C095's next authenticated pass
    /// clears it.
    var languagePendingSync: Bool { get set }

    /// Whether SCR-PI-005's rationale has been shown. The *rationale*, not the grant: iOS owns the
    /// grant and answers it through `CLLocationManager.authorizationStatus`, and a second copy of
    /// that here would be a stale one the first time somebody changed it in Settings.
    var locationRationaleAcknowledged: Bool { get set }

    /// C101's default payment rail (US-22.4). A string rather than a typed enum because the
    /// contract's own enum cannot express what this app offers — see the class documentation.
    var defaultPaymentMethod: String? { get set }

    /// SCR-PI-015a's remembered choice — `CallType.wire`, or `nil` before the first call (Δ C098).
    ///
    /// The cell says the sheet *"remembers last choice"*, and it has to **outlive the ride**: a
    /// passenger who always calls normally should not be asked again on their next trip. A wire
    /// string rather than a `CallType` for ``defaultPaymentMethod``'s reason — a value a later build
    /// wrote and this one has never heard of reads as *no preference* rather than as a crash.
    var lastCallType: String? { get set }

    /// Whether US-26.5's *"your number is visible to the other party"* notice has been shown
    /// (Δ C098).
    ///
    /// **Once, and only before a direct dial.** AL-48 withdrew masking outright, so this disclosure
    /// is the transparency that replaced it; showing it before a *free* call would be warning about
    /// something that is not happening, which is how people learn to dismiss disclosures.
    var callNumberNoticeShown: Bool { get set }
}

extension AppPreferences {

    /// Whether SCR-PI-002 has been answered — *"first-launch only"* (D2' §A). Δ C095.
    ///
    /// **Derived rather than stored**, which is the same call `AppPreferences.kt` makes. A separate
    /// flag would be a second fact that can disagree with the first: a passenger with a stored
    /// language and a `false` flag would be sent back to a screen that has nothing left to ask them,
    /// and one with no language and a `true` flag would meet the login screen in whatever locale the
    /// handset happens to be set to — which for most users here is not one of the three (AL-26).
    ///
    /// This is why ``OnboardingModel/finish()`` writes the language **unconditionally**, including
    /// when the passenger accepted the Sinhala default without touching a box: accepting it *is*
    /// answering the screen.
    var firstRunComplete: Bool { language != nil }
}

/// ``AppPreferences`` over `UserDefaults`.
///
/// The standard suite rather than a named one: this app has no extension and no app group, and a
/// suite that is not `.standard` is one more thing that has to be right for `AppleLanguages` — which
/// ``PassengerLocale`` writes — to be read by the system on the next launch.
final class UserDefaultsAppPreferences: AppPreferences {

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    var language: Language? {
        // `Language.companion.fromWire` rather than a table here: a value written by a later build
        // answers `nil`, which re-asks SCR-PI-002 — better than silently drawing the app in a
        // language the passenger did not pick, and better than a second copy of the three codes.
        get { defaults.string(forKey: Keys.language).flatMap { Language.companion.fromWire(wire: $0) } }
        set { defaults.set(newValue?.wire, forKey: Keys.language) }
    }

    var languagePendingSync: Bool {
        get { defaults.bool(forKey: Keys.pendingSync) }
        set { defaults.set(newValue, forKey: Keys.pendingSync) }
    }

    var locationRationaleAcknowledged: Bool {
        get { defaults.bool(forKey: Keys.locationRationale) }
        set { defaults.set(newValue, forKey: Keys.locationRationale) }
    }

    var defaultPaymentMethod: String? {
        get { defaults.string(forKey: Keys.defaultPayment) }
        set { defaults.set(newValue, forKey: Keys.defaultPayment) }
    }

    var lastCallType: String? {
        get { defaults.string(forKey: Keys.lastCallType) }
        set { defaults.set(newValue, forKey: Keys.lastCallType) }
    }

    var callNumberNoticeShown: Bool {
        get { defaults.bool(forKey: Keys.callNumberNotice) }
        set { defaults.set(newValue, forKey: Keys.callNumberNotice) }
    }

    private enum Keys {
        static let language = "language"
        static let pendingSync = "language_pending_sync"
        static let locationRationale = "location_rationale_acknowledged"
        static let defaultPayment = "default_payment_method"
        static let lastCallType = "last_call_type"
        static let callNumberNotice = "call_number_notice_shown"
    }
}
